profiling/src/internal/backup_sample_types.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::api::{Period, ValueType};
+use std::mem::ManuallyDrop;
+
+/// In order to reset the profile, it needs the sample types. However, it's
+/// inconvenient to reconstruct this because of differences in api and internal
+/// layouts. This struct serves as a backup of the original sample types, so
+/// we can just re-use them as needed. It does a single allocation for
+/// everything from the global allocator.
+pub struct BackupSampleTypesAndPeriod {
+    /// The static bounds are lies, they are tied to the `storage`. Only pass
+    /// out references with the same lifetime as &self.
+    sample_types: ManuallyDrop<Box<[ValueType<'static>]>>,
+    /// Despite our manual drop order, it is still recommended to order the
+    /// fields in the order things should be dropped.
+    storage: ManuallyDrop<Vec<(String, String)>>,
+    period: Option<((String, String), i64)>,
+}
+
+impl BackupSampleTypesAndPeriod {
+    /// Creates a backup of the value types.
+    pub fn new(sample_types: &[ValueType], period: Option<Period>) -> Self {
+        let mut storage = ManuallyDrop::new(Vec::with_capacity(sample_types.len()));
+        for sample_type in sample_types {
+            let item = (
+                String::from(sample_type.r#type),
+                String::from(sample_type.unit),
+            );
+            storage.push(item);
+        }
+
+        let mut new_sample_types = Vec::new();
+        new_sample_types.reserve_exact(sample_types.len());
+        for sample_type in storage.iter() {
+            // SAFETY: transforming the lifetime here is _NOT_ safe! We are
+            // appeasing the borrow-checker. Only expose it to the outside
+            // with a lifetime bound to &self!
+            let value_type = unsafe {
+                ValueType {
+                    r#type: core::mem::transmute(sample_type.0.as_str()),
+                    unit: core::mem::transmute(sample_type.1.as_str()),
+                }
+            };
+            new_sample_types.push(value_type);
```

Comment:
So, I'm not quite sure what's the advantage here between having the separate storage and sample_types. Specifically, the Rust docs say that a `String` always allocates its data on the heap, so assuming that's correct, I don't get what we're gaining from not having the sample types own the strings directly?

(This is probably me missing some context on some Rust limitation, as usual)

Reply:
No, that makes sense. Originally I had a smarter implementation which reduced allocations but miri didn't like it. I usually can figure out miri errors or at least understand them, and on that one I didn't. So I did a simpler thing.

But taking a step back, no, you are right. If I can't reduce this to like 1 allocation for the whole thing, may as well just make simple structs that own `String`s instead.